File: leetcode/editor/cn/405_数字转换为十六进制数_test.go
package leetcode

import (
	"github.com/stretchr/testify/assert"
	"testing"
)

func Test_givenNormal_when405_thenSuccess(t *testing.T) {
	assert.Equal(t, "1a", toHex(26))
	assert.Equal(t, "ffffffff", toHex(-1))
	assert.Equal(t, "ffff0001", toHex(-0xffff))
}

/**
边界条件 0
*/
func Test_givenFailedCase1_when405_thenSuccess(t *testing.T) {
	assert.Equal(t, "0", toHex(0))
}

//leetcode submit region begin(Prohibit modification and deletion)
func toHex(num int) string {
	if num == 0 {
		return "0"
	}
	ans := ""
	hexTab := []string{"0", "1", "2", "3", "4", "5", "6", "7",
		"8", "9", "a", "b", "c", "d", "e", "f"}
	// TODO GO -1 = 0xffffffffffffffff 且 Go 中无无符号右移
	mask, n := uint32(0x0f), uint32(num)
	for n != 0 {
		ans = hexTab[(n&mask)] + ans
		n >>= 4
	}
	return ans
}

//leetcode submit region end(Prohibit modification and deletion)

File: leetcode/editor/cn/405_数字转换为十六进制数_test.py
import unittest


class TestSolution(unittest.TestCase):

    def setUp(self) -> None:
        self.solution = Solution()

    def test_givenNormal_when_thenSuccess(self):
        self.assertEqual('1a', self.solution.toHex(26))
        self.assertEqual('0', self.solution.toHex(0))
        self.assertEqual('ffffffff', self.solution.toHex(-1))


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def toHex(self, num: int) -> str:
        if num == 0:
            return '0'
        hex_tab = '0123456789abcdef'
        ans = ''
        for i in range(7, -1, -1):
            ans += hex_tab[num >> (4 * i) & 0x0f]
        # 移除前导 0
        return ans.lstrip('0')

    # 此解法是错误的
    def toHexError(self, num: int) -> str:
        if num == 0:
            return '0'
        hex_tab = '0123456789abcdef'
        ans = ''
        while num:
            # TODO -1 >> 4 的结果依旧是 -1
            ans = hex_tab[num & 0x0f] + ans
            num >>= 4
        return ans


# leetcode submit region end(Prohibit modification and deletion)


if __name__ == '__main__':
    unittest.main()
